### PR TITLE
Ensure required values are set when changing `storage_type`.

### LINKED ISSUE #28589: [Bug]: "aws_db_instance", "storage_type" (io1 > gp3) can't be changed if "allocated_storage" is under 400 GB
Relates #28847 
Relates #33512 
Relates #33529 
Relates #36648
Relates #37257 

### Terraform Core Version

1.0.9

### AWS Provider Version

4.48.0

### Affected Resource(s)

aws_db_instance

### Expected Behavior

"storage_type" can be changed from "io1" to "gp3"

### Actual Behavior

"storage_type" can't be changed (io1 > gp3) as far as you have under 400 GB disk

if you do not specify "iops":

Error: updating RDS DB Instance (delme2): operation error RDS: ModifyDBInstance, https response error StatusCode: 400, RequestID: 20ed3bd9-1794-4451-a314-6eff87d6e851, api error InvalidParameterCombination: You must specify both the storage size and iops when modifying the storage size or iops on a DB instance that has iops.

if you do specify "iops":

`Error: updating RDS DB Instance (inapp-stage-delme2): operation error RDS: ModifyDBInstance, https response error StatusCode: 400, RequestID: 44cad8d2-dea0-4c77-9979-2314cc8d528a, api error InvalidParameterCombination: You can't specify IOPS or storage throughput for engine postgres and a storage size less than 400.`

### Relevant Error/Panic Output Snippet

_No response_

### Terraform Configuration Files

```hcl
resource "aws_db_instance" "delme2" {
  identifier = "delme2"

  engine            = "postgres"
  engine_version    = "13.7"
  instance_class    = "db.t4g.small"
  allocated_storage = 100
  storage_encrypted = true
  apply_immediately = true
  storage_type      = "io1"
  iops              = 1000

  iam_database_authentication_enabled = true

  username = "delme_db_admin"
  password = "P@ssw0rd"
  port     = "5432"

  parameter_group_name = "default.postgres12"

  deletion_protection = false
}
```

### Steps to Reproduce

1. create a new RDS instance ( storage_type = "io1", allocated_storage = 100, iops = 1000 )
2. change storage_type to "gp3" with or without specifying "iops" value

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by fatmcgav**
### Description
As observed in #28589, if changing the `storage_type` additional fields need to be set if required.

For example, `gp3` storage requires that the `allocated_storage` and `iops` params are supplied if the volume size is `>=400G`.

Added a new test case to try and exercise this behaviour, and also tweaked a couple of existing tests to exercise the `gp3` tiering behaviour. **N.B** I've not been able to run these acceptance tests locally yet.

### Relations

Fixes #28589

### Output from Acceptance Testing

N/A

### CODE REVIEW

**Comment by ewbankkit** [comment]
Relates https://github.com/hashicorp/terraform-provider-aws/pull/37257.
Relates https://github.com/hashicorp/terraform-provider-aws/pull/33529.

File: internal/service/rds/instance_test.go [hunk 1]
```diff
@@ -3727,79 +3727,6 @@ func TestAccRDSInstance_MonitoringRoleARN_removedToEnabled(t *testing.T) {
 	})
 }
 
-// Regression test for https://github.com/hashicorp/terraform/issues/3760 .
```
Comment (YakDriver) [inline internal/service/rds/instance_test.go:3730]: These are just moved to be near the other storage-related tests

**Review by YakDriver — DISMISSED** [review]
LGTM 🎉 

```
% make t T=TestAccRDSInstance_Storage K=rds         
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.3 test ./internal/service/rds/... -v -count 1 -parallel 20 -run='TestAccRDSInstance_Storage'  -timeout 360m
2024/11/22 13:57:21 Initializing Terraform AWS Provider...
=== RUN   TestAccRDSInstance_Storage_maxAllocated
=== PAUSE TestAccRDSInstance_Storage_maxAllocated
=== RUN   TestAccRDSInstance_Storage_gp3MySQL
=== PAUSE TestAccRDSInstance_Storage_gp3MySQL
=== RUN   TestAccRDSInstance_Storage_gp3Postgres
=== PAUSE TestAccRDSInstance_Storage_gp3Postgres
=== RUN   TestAccRDSInstance_Storage_gp3SQLServer
=== PAUSE TestAccRDSInstance_Storage_gp3SQLServer
=== RUN   TestAccRDSInstance_Storage_changeThroughput
=== PAUSE TestAccRDSInstance_Storage_changeThroughput
=== RUN   TestAccRDSInstance_Storage_changeIOPSThroughput
=== PAUSE TestAccRDSInstance_Storage_changeIOPSThroughput
=== RUN   TestAccRDSInstance_Storage_changeIOPSGP3
=== PAUSE TestAccRDSInstance_Storage_changeIOPSGP3
=== RUN   TestAccRDSInstance_Storage_throughputSSE
=== PAUSE TestAccRDSInstance_Storage_throughputSSE
=== RUN   TestAccRDSInstance_Storage_postgres
=== PAUSE TestAccRDSInstance_Storage_postgres
=== RUN   TestAccRDSInstance_Storage_changeIOPSio1
=== PAUSE TestAccRDSInstance_Storage_changeIOPSio1
=== RUN   TestAccRDSInstance_Storage_changeIOPSio2
=== PAUSE TestAccRDSInstance_Storage_changeIOPSio2
=== CONT  TestAccRDSInstance_Storage_maxAllocated
=== CONT  TestAccRDSInstance_Storage_changeIOPSGP3
=== CONT  TestAccRDSInstance_Storage_gp3SQLServer
=== CONT  TestAccRDSInstance_Storage_gp3Postgres
=== CONT  TestAccRDSInstance_Storage_changeIOPSio2
=== CONT  TestAccRDSInstance_Storage_postgres
=== CONT  TestAccRDSInstance_Storage_throughputSSE
=== CONT  TestAccRDSInstance_Storage_changeIOPSThroughput
=== CONT  TestAccRDSInstance_Storage_changeThroughput
=== CONT  TestAccRDSInstance_Storage_gp3MySQL
=== CONT  TestAccRDSInstance_Storage_changeIOPSio1
--- PASS: TestAccRDSInstance_Storage_changeIOPSGP3 (538.54s)
--- PASS: TestAccRDSInstance_Storage_changeIOPSio1 (650.44s)
--- PASS: TestAccRDSInstance_Storage_changeIOPSio2 (660.37s)
--- PASS: TestAccRDSInstance_Storage_changeThroughput (660.92s)
--- PASS: TestAccRDSInstance_Storage_changeIOPSThroughput (670.59s)
--- PASS: TestAccRDSInstance_Storage_maxAllocated (684.31s)
--- PASS: TestAccRDSInstance_Storage_gp3MySQL (726.56s)
--- PASS: TestAccRDSInstance_Storage_gp3Postgres (785.79s)
--- PASS: TestAccRDSInstance_Storage_postgres (836.50s)
--- PASS: TestAccRDSInstance_Storage_throughputSSE (1103.82s)
--- PASS: TestAccRDSInstance_Storage_gp3SQLServer (1289.33s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/rds	1293.310s
```

File: .changelog/37257.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+```release-note:bug
+resource/aws_db_instance: When changing from io1/io2 to gp3, fix bug causing error `InvalidParameterCombination: You must specify both the storage size and iops when modifying the storage size or iops on a DB instance that has iops`
```
Comment (ewbankkit) [inline .changelog/37257.txt:2]:
```suggestion
resource/aws_db_instance: When changing `storage_type` from `io1` or `io2` to `gp3`, fix bug causing error `InvalidParameterCombination: You must specify both the storage size and iops when modifying the storage size or iops on a DB instance that has iops`
```

File: .changelog/28847.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+```release-note:bug
+resource/aws_db_instance: When changing to gp3 larger than allocated storage threshold, fix bug causing error `InvalidParameterCombination: You must specify both the storage size and iops when modifying the storage size or iops on a DB instance that has iops`
```
Comment (ewbankkit) [inline .changelog/28847.txt:2]:
```suggestion
resource/aws_db_instance: When changing a `gp3` volume's `allocated_storage` to a value larger than the [threshold value for `engine`](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_Storage.html#gp3-storage), fix bug causing error `InvalidParameterCombination: You must specify both the storage size and iops when modifying the storage size or iops on a DB instance that has iops`
```

**Review by YakDriver — APPROVED** [review]

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

**Pull request merged**